File: binary_tree_maximum_path_sum.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root):
        """
        计算二叉树的最大路径和
        
        Args:
            root: 二叉树的根节点
            
        Returns:
            int: 最大路径和
        """
        # 使用一个列表来存储全局最大值，因为我们需要在递归过程中更新它
        self.max_sum = float('-inf')
        
        def max_gain(node):
            """
            计算以当前节点为根的子树的单边最大路径和
            同时更新全局最大路径和
            
            Args:
                node: 当前节点
                
            Returns:
                int: 以当前节点为起点的单边最大路径和
            """
            if not node:
                return 0
            
            # 递归计算左右子树的单边最大路径和
            left_gain = max(max_gain(node.left), 0)
            right_gain = max(max_gain(node.right), 0)
            
            # 计算包含当前节点的路径和（可以包含左右子树）
            price_newpath = node.val + left_gain + right_gain
            
            # 更新全局最大路径和
            self.max_sum = max(self.max_sum, price_newpath)
            
            # 返回以当前节点为起点的单边最大路径和
            # 注意：这里只能选择左子树或右子树中的一个，不能同时选择两个
            return node.val + max(left_gain, right_gain)
        
        max_gain(root)
        return self.max_sum

def build_tree_from_list(values):
    """
    从列表构建二叉树
    
    Args:
        values: 列表形式的二叉树表示
        
    Returns:
        TreeNode: 二叉树的根节点
    """
    if not values:
        return None
    
    root = TreeNode(values[0])
    queue = [root]
    i = 1
    
    while queue and i < len(values):
        node = queue.pop(0)
        
        # 添加左子节点
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        
        # 添加右子节点
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    
    return root

def print_tree(root, level=0, prefix="Root: "):
    """
    打印二叉树结构（用于调试）
    """
    if root is None:
        return
    
    print("  " * level + prefix + str(root.val))
    if root.left:
        print_tree(root.left, level + 1, "L--- ")
    if root.right:
        print_tree(root.right, level + 1, "R--- ")

# 测试用例
if __name__ == "__main__":
    # 示例 1: [1,2,3]
    print("示例 1:")
    values1 = [1, 2, 3]
    root1 = build_tree_from_list(values1)
    print("输入:", values1)
    print("二叉树结构:")
    print_tree(root1)
    
    solution = Solution()
    result1 = solution.maxPathSum(root1)
    print("输出:", result1)
    print("解释: 最优路径是 2 -> 1 -> 3，路径和为 2 + 1 + 3 = 6")
    print()
    
    # 更多测试用例
    print("更多测试用例:")
    
    # 测试用例 2: [-10,9,20,None,None,15,7]
    values2 = [-10, 9, 20, None, None, 15, 7]
    root2 = build_tree_from_list(values2)
    print("输入:", values2)
    result2 = solution.maxPathSum(root2)
    print("输出:", result2)
    print()
    
    # 测试用例 3: [2,-1]
    values3 = [2, -1]
    root3 = build_tree_from_list(values3)
    print("输入:", values3)
    result3 = solution.maxPathSum(root3)
    print("输出:", result3)
    print()
    
    # 测试用例 4: [1]
    values4 = [1]
    root4 = build_tree_from_list(values4)
    print("输入:", values4)
    result4 = solution.maxPathSum(root4)
    print("输出:", result4)

File: BinaryTreeMaximumPathSum.java
/**
 * 二叉树最大路径和
 * 
 * 题目描述：在二叉树中，路径被定义为一条节点序列，序列中每对相邻节点之间都存在一条边，
 * 同一个节点在一条路径序列中至多出现一次，该路径至少包含一个节点，且不一定经过根节点。
 * 路径和是路径中各节点值的总和。给定一个二叉树的根节点root，要求返回其最大路径和。
 * 
 * 示例1：
 * 输入：root = [1,2,3]
 * 输出：6
 * 解释：最优路径是2 -> 1 -> 3，路径和为2 + 1 + 3 = 6
 */

// 定义二叉树节点
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class BinaryTreeMaximumPathSum {
    
    private int maxSum = Integer.MIN_VALUE;
    
    /**
     * 计算二叉树的最大路径和
     * @param root 二叉树的根节点
     * @return 最大路径和
     */
    public int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        maxGain(root);
        return maxSum;
    }
    
    /**
     * 递归计算以当前节点为根的最大贡献值
     * @param node 当前节点
     * @return 以当前节点为根的最大贡献值
     */
    private int maxGain(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        // 递归计算左子树的最大贡献值
        int leftGain = Math.max(maxGain(node.left), 0);
        
        // 递归计算右子树的最大贡献值
        int rightGain = Math.max(maxGain(node.right), 0);
        
        // 计算以当前节点为根的新路径和
        int priceNewPath = node.val + leftGain + rightGain;
        
        // 更新全局最大路径和
        maxSum = Math.max(maxSum, priceNewPath);
        
        // 返回以当前节点为根的最大贡献值（只能选择左子树或右子树之一）
        return node.val + Math.max(leftGain, rightGain);
    }
    
    /**
     * 测试方法
     */
    public static void main(String[] args) {
        BinaryTreeMaximumPathSum solution = new BinaryTreeMaximumPathSum();
        
        // 测试示例1: [1,2,3]
        TreeNode root1 = new TreeNode(1);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(3);
        
        System.out.println("示例1 - 输入: [1,2,3]");
        System.out.println("输出: " + solution.maxPathSum(root1));
        System.out.println("期望: 6");
        System.out.println();
        
        // 测试示例2: [-10,9,20,null,null,15,7]
        TreeNode root2 = new TreeNode(-10);
        root2.left = new TreeNode(9);
        root2.right = new TreeNode(20);
        root2.right.left = new TreeNode(15);
        root2.right.right = new TreeNode(7);
        
        System.out.println("示例2 - 输入: [-10,9,20,null,null,15,7]");
        System.out.println("输出: " + solution.maxPathSum(root2));
        System.out.println("期望: 42");
        System.out.println();
        
        // 测试示例3: [2,-1]
        TreeNode root3 = new TreeNode(2);
        root3.left = new TreeNode(-1);
        
        System.out.println("示例3 - 输入: [2,-1]");
        System.out.println("输出: " + solution.maxPathSum(root3));
        System.out.println("期望: 2");
        System.out.println();
        
        // 测试示例4: [1]
        TreeNode root4 = new TreeNode(1);
        
        System.out.println("示例4 - 输入: [1]");
        System.out.println("输出: " + solution.maxPathSum(root4));
        System.out.println("期望: 1");
    }
}